syntax = "proto2";
package mltools;

// represent the feature value of specific feature group id
// this is widely used in VW to incorporate multiple feature source
// for each example
message SlotInfo {
    enum Format {
        DENSE = 1;
        SPARSE = 2;
        SPARSE_BINARY = 3;
    }

    optional Format format = 1;
    optional int32 id = 2;
    optional uint64 min_key = 3 [default=0xFFFFFFFFFFFFFFFF];
    optional uint64 max_key = 4;
    optional uint64 nnz_ele = 5;
    optional uint64 nnz_ex = 6;
}

// each sample will have multiple slot, with different features
// each slot can have different set of keys
message ExampleInfo {
    repeated SlotInfo slot = 1;
    optional uint64 num_ex = 2;
}

message Slot {
    optional int32 id = 1;
    repeated uint64 key = 2 [packed=true];
    repeated float val = 3 [packed=true];
    optional float wgt = 4 [default=1.0];
}

message Example {
    repeated Slot slot = 1;
    optional float wgt = 2 [default=1.0];
}
